src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+         // As this is a newly allocated fragment being tracked, it should
+         // not be found elsewhere in the tracker array.
+         platform_assert((frag->shm_frag_addr != addr),
+                         "Error! Newly allocated large memory fragment at %p"
+                         " is already tracked at slot %d."
+                         " Fragment is allocated to PID=%d, to tid=%lu,"
+                         " and is %s (freed by PID=%d, by tid=%lu)\n.",
+                         addr,
+                         fctr,
+                         frag->shm_frag_allocated_to_pid,
+                         frag->shm_frag_allocated_to_tid,
+                         (frag->shm_frag_freed_by_pid ? "free" : "in use"),
+                         frag->shm_frag_freed_by_pid,
+                         frag->shm_frag_freed_by_tid);
+         continue;
+      }
+
+      // We found a free slot. Track our memory fragment at fctr'th slot.
+      shm->shm_num_frags_tracked++;
+      shm->shm_num_frags_inuse++;
+      shm->shm_used_by_large_frags += size;
+      if (shm->shm_num_frags_inuse > shm->shm_num_frags_inuse_HWM) {
+         shm->shm_num_frags_inuse_HWM = shm->shm_num_frags_inuse;
+      }
+
+      // We should really assert that the other fields are zero, but for now
+      // re-init this fragment tracker.
+      memset(frag, 0, sizeof(*frag));
+      frag->shm_frag_addr = addr;
+      frag->shm_frag_size = size;
+
+      frag->shm_frag_allocated_to_pid = getpid();
+      frag->shm_frag_allocated_to_tid = platform_get_tid();
+
+      // The freed_by_pid/freed_by_tid == 0 means fragment is still allocated.
+
+      // Track highest address of large-fragment that is being tracked.
+      if (shm->shm_large_frag_hip < addr) {
+         shm->shm_large_frag_hip = addr;
+      }
+      break;
+   }
+   shm_unlock_mem_frags(shm);
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shm_track_free() - See if this memory fragment being freed is
+ * already being tracked. If so, it's a large fragment allocation, which can be
+ * re-cycled after this free. Do the book-keeping accordingly to record that
+ * this large-fragment is no longer in-use and can be recycled.
+ * -----------------------------------------------------------------------------
+ */
+static void
+platform_shm_track_free(shmem_info *shm,
+                        void       *addr,
+                        const char *objname,
+                        const char *func,
+                        const char *file,
+                        const int   lineno)
+{
+   shm_lock_mem_frags(shm);
+   shm->shm_nfrees++;
+
+   // If we are freeing a fragment beyond the high-address of all
+   // large fragments tracked, then this is certainly not a large
+   // fragment. So, no further need to see if it's a tracked fragment.
+   if (addr > shm->shm_large_frag_hip) {
+      shm->shm_nf_search_skipped++; // Track # of optimizations done
```

Comment:
This is an important optimization to avoid searching needlessly in loop below when small fragments are freed.

Recall that in this version we don't have a real free().